### PR TITLE
Handle empty and missing response bodies.

### LINKED ISSUE #242: Chucker 3.1.2 re-writing the Server Response; throws IOException when used as a networkInterceptor
## :writing_hand: Describe the bug
After updating to 3.1.2 (from 3.0.1) every request fails with an `IOException`:
```
java.io.IOException: ID1ID2: actual 0x00007b22 != expected 0x00001f8b
System.err:     at okio.GzipSource.checkEqual(GzipSource.java:205)
System.err:     at okio.GzipSource.consumeHeader(GzipSource.java:120)
System.err:     at okio.GzipSource.read(GzipSource.java:73)
System.err:     at okio.RealBufferedSource.read(RealBufferedSource.java:51)
System.err:     at okio.ForwardingSource.read(ForwardingSource.java:35)
System.err:     at retrofit2.OkHttpCall$ExceptionCatchingResponseBody$1.read(OkHttpCall.java:288)
System.err:     at okio.RealBufferedSource.select(RealBufferedSource.java:100)
```

Chucker `3.0.1` returned the original Response. `3.1.x` seems to process the response and returns this "processed Response" (see [here](https://github.com/ChuckerTeam/chucker/blob/3.1.2/library/src/main/java/com/chuckerteam/chucker/api/ChuckerInterceptor.kt#L68)).

I see the need of processing the Response and doing stuff with it. But the Interceptor has to return the original response at the end, since Chucker is only some kind of debugging ouput, which should not change the Request/Response in any way.

## :bomb: Steps to reproduce
- Add `ChuckerInterceptor` as a `networkInterceptor`
- I think the Request has to be gzipped

## :wrench: Expected behavior
- Chucker should not alter the Server Response

## :iphone: Tech info
Chucker version: `3.1.2`
OkHttp version: `3.14.6`

## :page_facing_up: Additional context
This only happens when the `ChuckerInterceptor` is added as a `networkInterceptor`.
The Notification shows the correct Request with the Response.

### PR DESCRIPTION
**Pull request by MiSikora**
## :page_facing_up: Context
There is an issue - #242.

## :pencil: Changes
I made `ChuckerInterceptor` gzip response back to the end end consumer in case it had to gunzip it. I also added tests for cases where `ChuckerInterceptor` is used as a netework interceptor.

This is rather a quick fix as it affects some people. The proper solution would be to split requests and responses streaming in order to allow Chucker to consume them separately from the end user.

## :no_entry_sign: Breaking
No.

## :hammer_and_wrench: How to test
You can add a `ChuckerInterceptor` as a network interceptor. It should not cause `IOException`s for gzipped responses. You can also check the 81accac0 commit and run tests. They shoud fail at that stage.

## :stopwatch: Next steps
No steps.

Closes #242

### CODE REVIEW

File: library/src/test/java/com/chuckerteam/chucker/api/ChuckerInterceptorTest.kt [hunk 1]
```diff
@@ -15,76 +14,120 @@ import okio.Buffer
 import okio.ByteString
 import okio.GzipSink
 import org.junit.Rule
-import org.junit.Test
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.EnumSource
 
 class ChuckerInterceptorTest {
+    enum class ClientFactory {
+        REGULAR {
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply library/src/test/java/com/chuckerteam/chucker/api/ChuckerInterceptorTest.kt]: Yeah, you are right. I forgot that they are called application interceptors.

**Comment by jannisveerkamp** [comment]
I still got an error with this PR for some Requests. Most of them work though.
```
HTTP FAILED: java.io.EOFException
System.err: java.io.EOFException
System.err:     at okio.RealBufferedSource.require(RealBufferedSource.kt:201)
System.err:     at okio.GzipSource.consumeHeader(GzipSource.kt:104)
System.err:     at okio.GzipSource.read(GzipSource.kt:62)
System.err:     at okio.Buffer.writeAll(Buffer.kt:1650)
System.err:     at com.chuckerteam.chucker.api.ChuckerInterceptor.processResponseBody(ChuckerInterceptor.kt:149)
System.err:     at com.chuckerteam.chucker.api.ChuckerInterceptor.processResponse(ChuckerInterceptor.kt:131)
System.err:     at com.chuckerteam.chucker.api.ChuckerInterceptor.intercept(ChuckerInterceptor.kt:63)
System.err:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:112)
```

**Comment by MiSikora** [comment]
@jannisveerkamp 
Can you tell more about the nature of failing requests? Method type, response code, response headers, response body, etc? Are all of them the same and only some of them ocasionally fail or is it only for some endpoints that you use?

From the stacktrace it looks like Gzip header bytes are missing in the body while there is a response header `Content-Encoding: gzip`.

**Comment by MiSikora** [comment]
@vbuberen @cortinico 
What they do in `OkHttp` before gunzipping is they check if there is actually a promise of a body.

https://github.com/square/okhttp/blob/1256f1371b028d45f3d71d1ee8ae282f3aad912e/okhttp/src/main/java/okhttp3/internal/http/BridgeInterceptor.kt#L90-L92

Maybe we should do something like that here as well and cover it with tests?

**Comment by jannisveerkamp** [comment]
Now it worked everythere when I retestet 🙈

**Comment by MiSikora** [comment]
No worries, fortunately I can reproduce it in tests. I just want feedback from owners first.

**Comment by jannisveerkamp** [comment]
I think it happened when I run into a Cache (304, not max-age)

**Comment by MiSikora** [comment]
Ok, that makes sense as well. I added some guarding for no or empty content.

File: library/src/test/java/com/chuckerteam/chucker/ChuckerInterceptorDelegate.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    maxContentLength: Long = 250000L,
+    headersToRedact: Set<String> = emptySet()
+) : Interceptor {
+    private val idGenerator = AtomicLong()
+    private val transactions = CopyOnWriteArrayList<HttpTransaction>()
+
+    private val mockContext = mockk<Context> {
+        every { getString(any()) } returns ""
+    }
+    private val mockCollector = mockk<ChuckerCollector>() {
```
Comment (cortinico) [inline library/src/test/java/com/chuckerteam/chucker/ChuckerInterceptorDelegate.kt]:
```suggestion
    private val mockCollector = mockk<ChuckerCollector> {
```

File: library/src/test/java/com/chuckerteam/chucker/api/ChuckerInterceptorTest.kt [hunk 3]
```diff
@@ -15,77 +13,174 @@ import okio.Buffer
...
+        val client = factory.create(chucker)
+        val responseBody = client.newCall(request).execute().readByteStringBody()!!
+
+        assertThat(responseBody.utf8()).isEqualTo("Hello, world!")
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = ClientFactory::class)
+    fun gzippedBody_withNoContent_isTransparentForChucker(factory: ClientFactory) {
+        server.enqueue(MockResponse().addHeader("Content-Encoding: gzip").setResponseCode(204))
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply library/src/test/java/com/chuckerteam/chucker/api/ChuckerInterceptorTest.kt]: Personally, I prefer this over constants in cases like this but I understand where are you coming from. I will change it. :)

**Comment by MiSikora** [comment]
> @MiSikora When I mentioned cloning I meant creating a copy using standard `Response.Builder` class. Your approach might work as well, but the only concern there is to be sure that we won't reintroduce #203 again somehow. It seems we won't, but anyway.

Yeah, I'm not sure here to be honest. I reverted the change of returning custom response because I wanted to undo the change I did in #226, write tests and then eventually migrate to build custom response to make sure that old tests still pass. What do you think?

**Comment by vbuberen** [comment]
I believe it should work and we are good to do. 
I just wanted to hear a confirmation from @jannisveerkamp about his issue being fixed.

**Comment by vbuberen** [comment]
@jannisveerkamp Could you please respond if this PR fixes the issue you opened?

**Comment by MohabTarek** [comment]
@jannisveerkamp @vbuberen Hi guys any updates about this PR to be merged soon ?

**Comment by MiSikora** [comment]
As the author of this PR I'd like to see #267 resolved first in either way - merged or closed. It will be easier for me to resolve conflicts on this branch in case #267 gets merged.

**Comment by cortinico** [comment]
> @MohabTarek As the author of this PR I'd like to see #267 resolved first in either way - merged or closed. It will be easier for me to resolve conflicts on this branch in case #267 gets merged.

Agree on this, let's land #267 first. Moreover, @MohabTarek you can grab an artifact from @MiSikora's fork from Jitpack if you're blocked by this.

**Comment by vbuberen** [comment]
@MiSikora I am quite lost here - are we good to merge this or you want to modify it further?

**Comment by MiSikora** [comment]
Yes, from my side it is ready. I just pushed 1 small commit with a fix for exceeding capacity in side channels.

**Comment by vbuberen** [comment]
Awesome, thanks for doing that big amount of work.

**Comment by jannisveerkamp** [comment]
I'm sorry. I wasn't available. I will test as soon as I find the time 😃

**Pull request merged**